### ISSUE TITLE
OT really works?

### ISSUE THREAD

**Issue by wjczf123** [CHENGY12/PLOT#6]
I reproduced the experiment and found that the learned T is uniform. 
Then I try to set T to a uniform matrix and found that the learned T via OT achieves similar performance compared to using a uniform matrix without learning.

**Comment by wjczf123** [comment]
@CHENGY12 @hit-Chris Can you answer my question?

**Comment by CHENGY12** [comment]
Hi, please check if the learning process is correctly setup. We have double checked that for different prompts, the learned T is significantly different. As visualized, different prompts always focus on different regions.  If taking the T as ones, it is equal to do the mean pool, whose performance is even lower than coop (57 in ImageNet).   Please let us know if you still have issues.

**Comment by wjczf123** [comment]
Thanks for your reply. I run your code on caltech-101 dataset. We set sim_op = torch.sum(0.0051 * sim, dim=(1,2)) (i.e., unlearned and uniform T) and achieves similar performance. The learned T seems uniform.

**Comment by wjczf123** [comment]
<img width="447" alt="c23775a3f6192ec37f9ed9c81ed9c48" src="https://github.com/CHENGY12/PLOT/assets/42801458/cd8e37b3-5ee0-49d6-a608-28c9006ccd22">

**Comment by CHENGY12** [comment]
Thank you for your valuable points! We've investigated the two instances, observing distinct regions to be highlighted. Particularly, we utilized the first test images from Caltech101 and Stanford Cars datasets, which appear to manifest more significant observations compared to the previously shared examples.

We acknowledge that while the absolute differences between various transport plans are relatively minuscule, such as 0.007 and 0.003, their relative differences are indeed substantial. Thus, to facilitate a more coherent understanding, we recommend employing visualization in conjunction with the original image to examine the highlighted regions. For your convenience, you can access our visualization code at [visualization](https://github.com/CHENGY12/PLOT/blob/main/visualization.py). 

Here are a couple of visualization samples for your reference:
![1551697395278_ pic](https://github.com/CHENGY12/PLOT/assets/17920133/aa4a88bd-d5d1-47f0-a5da-07f6dda27a8e)
![1561697395569_ pic](https://github.com/CHENGY12/PLOT/assets/17920133/1836e2c5-61aa-4508-9708-d5879981b41b)

Regarding quantitative results, we concur that the performance demonstrates variances attributed to different environments and inherent randomness. Therefore, we suggest a comprehensive comparison using the average results across all datasets and seeds to ensure robustness and reliability in our findings.

As a side note, my responsiveness to GitHub issues might be slightly delayed. For a more prompt discussion, please feel free to reach out to me directly via email at [guangyichen1994@gmail.com](mailto:guangyichen1994@gmail.com).

Looking forward to our further discussions.

**Comment by wjczf123** [comment]
Thanks for your reply. I have another question about OT.
OT seems to be a non-parameterized attention mechanism.
Does it have any advantage compared to the attention mechanism?

**Comment by CHENGY12** [comment]
Good question!  
1)  From the technical side, attention is directly based on similarity, while OT considers the whole assignment between the distributions.  It means the ``similarity'' in OT will involve competition, not direct sum In the attention. For example, if similarities between one query and two galleries are 5.3 and 4.7. Then when considering the whole distribution assignment, the plan of OT may keep one is 1 while the other is 0. 
2) OT works well when the concepts are clear in our implementation (assuming the discrete distribution ).
3) Attention learning needs more data and may learn the shortcut [CAL](https://arxiv.org/abs/2108.08728), given the few-shot prompt learning setting, it may be difficult to learn well.  
4) Attention has trainable parameters, which is not suitable for the field of prompt learning scenarios.

**Comment by wjczf123** [comment]
Thanks for your reply. I got it. Hope you have a good day.
